newclient/client.go

```diff
@@ -0,0 +1,245 @@
+		}
+		c.pairingURL = pairing
+		return nil
+	}
+}
+
+// The WithRealmManagementURL function allows to specify an
+// RealmManagement URL different from the standard one (e.g. http://localhost:4003).
+// This is not recommendend in production.
+func WithRealmManagementURL(realmManagementURL string) clientOption {
+	return func(c *Client) error {
+		realmManagement, err := url.Parse(realmManagementURL)
+		if err != nil {
+			return err
+		}
+		c.realmManagementURL = realmManagement
+		return nil
+	}
+}
+
+// The WithBaseURL function allows to specify the Astarte
+// base URL (e.g. api.your-astarte.org)
+func WithBaseURL(baseURL string) clientOption {
+	return func(c *Client) error {
+		base, err := url.Parse(baseURL)
+		if err != nil {
+			return err
+		}
+		c.baseURL = base
+
+		return nil
+	}
+}
+
+// The WithHTTPClient function allows to specify an httpClient
+// with custom options, e.g. different timeout, or skipTLSVerify
+func WithHTTPClient(httpClient *http.Client) clientOption {
+	return func(c *Client) error {
+		c.httpClient = httpClient
+		return nil
+	}
+}
+
+// The WithToken function allows to specify a JWT
+// token that the client will use to interact with Astarte.
+func WithToken(token string) clientOption {
+	return func(c *Client) error {
+		c.token = token
+		return nil
+	}
+}
+
+// The WithUserAgent function allows to specify the User Agent
+// that the client will use when making http requests.
+func WithUserAgent(userAgent string) clientOption {
+	return func(c *Client) error {
+		c.userAgent = userAgent
+		return nil
+	}
+}
+
+// The WithPrivateKey function allows to specify a realm private key,
+// used internally to generate a valid JWT token to all Astarte APIs with no expiry.
+// The client will use that token to interact with Astarte.
+func WithPrivateKey[T privateKeyProvider](privateKey T) clientOption {
+	return WithPrivateKeyWithTTL(privateKey, 0)
+}
+
+// The WithPrivateKey function allows to specify a realm private key,
+// used internally to generate a valid JWT token to all Astarte APIs with a specified expiry (in seconds).
+// The client will use that token to interact with Astarte.
+func WithPrivateKeyWithTTL[T privateKeyProvider](privateKey T, ttlSeconds int64) clientOption {
+	// Add all types
+	servicesAndClaims := map[misc.AstarteService][]string{
+		misc.AppEngine:       {},
+		misc.Channels:        {},
+		misc.Flow:            {},
+		misc.Housekeeping:    {},
+		misc.Pairing:         {},
+		misc.RealmManagement: {},
+	}
+	return WithPrivateKeyWithClaimsWithTTL(privateKey, servicesAndClaims, 0)
+}
+
+// The WithPrivateKey function allows to specify a realm private key,
+// used internally to generate a valid JWT token with a given set of Astarte claims and
+// a specified expiry (in seconds).
+// The client will use that token to interact with Astarte.
+func WithPrivateKeyWithClaimsWithTTL[T privateKeyProvider](privateKey T, claims map[misc.AstarteService][]string, ttlSeconds int64) clientOption {
```

Comment:
Instead of piling up `PrivateKey` + `Claims` + `Expiry` adding one piece at a time, why don't you provide the 3 functions separately so they can be composed?
Also, similar to the problem @szakhlypa encountered in the Elixir client, are we sure we want to expose the `expiry` property to the end user and we want to generate tokens when the client is created? This will inevitably lead to clients that have expired JWTs in them. @szakhlypa how were you thinking to approach this during the next Elixir client iteration?

Reply:
After IRL discussion, the following points have been made:
- The client will generate a JWT for every request
- Since a private key is provided, it makes no sense at the moment to specify a restricted set of claims when generating the JWT
- An expiry must always be set; the default value will be 5 minutes but can be limited further by the user